keras_cv/models/__internal__/darknet_utils.py

```diff
@@ -0,0 +1,398 @@
+        activation="silu",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.filters = filters
+        self.num_bottlenecks = num_bottlenecks
+        self.add_residual = add_residual
+        self.use_depthwise = use_depthwise
+        self.activation = activation
+
+        hidden_channels = filters // 2
+        ConvBlock = DarknetConvBlockDepthwise if use_depthwise else DarknetConvBlock
+
+        self.darknet_conv1 = DarknetConvBlock(
+            hidden_channels,
+            kernel_size=1,
+            strides=1,
+            activation=activation,
+        )
+
+        self.darknet_conv2 = DarknetConvBlock(
+            hidden_channels,
+            kernel_size=1,
+            strides=1,
+            activation=activation,
+        )
+
+        # repeat bottlenecks num_bottleneck times
+        self.bottleneck_convs = []
+        for _ in range(num_bottlenecks):
+            self.bottleneck_convs.append(
+                DarknetConvBlock(
+                    hidden_channels,
+                    kernel_size=1,
+                    strides=1,
+                    activation=activation,
+                )
+            )
+
+            self.bottleneck_convs.append(
+                ConvBlock(
+                    hidden_channels,
+                    kernel_size=3,
+                    strides=1,
+                    activation=activation,
+                )
+            )
+
+        self.add = layers.Add()
+        self.concatenate = layers.Concatenate()
+
+        self.darknet_conv3 = DarknetConvBlock(
+            filters, kernel_size=1, strides=1, activation=activation
+        )
+
+    def call(self, x):
+        x1 = self.darknet_conv1(x)
+        x2 = self.darknet_conv2(x)
+
+        for i in range(self.num_bottlenecks):
+            residual = x1
+            x1 = self.bottleneck_convs[2 * i](x1)
+            x1 = self.bottleneck_convs[2 * i + 1](x1)
+
+            if self.add_residual:
+                x1 = self.add([residual, x1])
+
+        x1 = self.concatenate([x1, x2])
+        x = self.darknet_conv3(x1)
+        return x
+
+    def get_config(self):
+        config = {
+            "filters": self.filters,
+            "num_bottlenecks": self.num_bottlenecks,
+            "add_residual": self.add_residual,
+            "use_depthwise": self.use_depthwise,
+            "activation": self.activation,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+
+def Focus(name=None):
+    """A block used in CSPDarknet to focus information into channels of the image.
+
+    If the dimensions of a batch input is (batch_size, width, height, channels), this
+    layer converts the image into size (batch_size, width/2, height/2, 4*channels).
+
+    Args:
+        name: the name for the lambda layer used in the block.
+
+    Returns:
+        a function that takes an input Tensor representing a Focus layer.
+    """
+
+    def apply(x):
```

Comment:
interesting, I wish I understood the purpose of this better. 

Maybe a docstring comment can serve to document this?

Reply:
This is a new layer added with the YoloV5 implementation. Supposed to make the network more efficient.

https://github.com/ultralytics/yolov5/discussions/3181